### ISSUE TITLE
expose manifest for file before writing?

### ISSUE THREAD

**Issue by deitch** [google/go-containerregistry#680]
When I download an image, once I have the manifest, I can calculate the size of the final download by looking at the layers and the config; the manifest has sizes of both of those.

However, the final `v1.tarball` format includes those _and_ the `manifest.json`, which is written to to the file [here](https://github.com/google/go-containerregistry/blob/6950943e71a1805ce590c2b5ee50d761c09acc02/pkg/v1/tarball/write.go#L166-L172). This makes it very difficult to calculate the file size in advance, which I might need for various types of validation, size calculations, etc. Sure, it is small, but it throws things off.

Would we be willing to expose this manifest, since it is easily to pre-calculate once you have the manifest in hand? 

Essentially, it would be taking the code and making it an exposed func , that is consumed by that same line of code. I already do it outside, but would not like to be so dependent on the internals.

Ideally, I would love a `v1tarball.CalculateSize(v1.Image) int64` which would do this, but maybe I am asking for too much? 

Happy to open a PR for either.

**Comment by ImJasonH** [comment]
`image.RawManifest()` returns the raw JSON bytes of the manifest, you can `len` that and add it to the sizes of all the layers to get the full size. If that's all you need that seems simple enough to write in your own code, unless I'm misunderstanding the issue.

**Comment by deitch** [comment]
That isn’t quite it. Look at the lines I added, that is another manifest that is added by the lib when the tar file is created. That one is not in the raw manifest. 

I stumbled across this because I do exactly what you suggested and was off by a few bytes (to give credit, one of my colleagues found it). 

Ideally, there would just be a size func that gives the whole expected size, but enough if I can get the size of that internal manifest to add to the calc you listed above.

**Comment by ImJasonH** [comment]
Oh I see, the issue is the tar header overhead. Is the overhead a constant number or are you seeing it fluctuate? Could you add some buffer to account for the overhead?

I'm reluctant to add this to the library if it's fairly straightforward to calculate/approximate.

**Comment by deitch** [comment]
Fluctuates. It depends on quite a bit. I tried to calculate it by copying the code and trying to replicate it (awful violation of DRY and coupling, guaranteed to break with first library change, I know), but even then missed something. 

Why the hesitancy to add something that doesn’t affect existing APIs?

**Comment by jonjohnsonjr** [comment]
I'm not sure that we can accurately know the resulting size of the tarball before without writing out the whole thing... can we? Unless you just care about the file sizes within the tarball and not the tarball itself? What's your ultimate goal?

In terms of structuring the code, we could do something stupid like we do in `logs`... https://github.com/google/go-containerregistry/blob/ba53fa10e72c7373ef8068e21c9f102f0c20a940/pkg/logs/logs.go#L38

... where we share all that code but _don't_ write out any files if we pass in `ioutil.Discard` as the `io.Writer`, just return a list of `tar.Header`s (or something).

One thing I'm a little worried about is that this makes it harder to work with [streaming layers](https://github.com/google/go-containerregistry/tree/master/pkg/v1/stream), but `tarball.Write` doesn't currently work with those right now either way, so it's not a huge deal... just something to consider.

**Comment by ImJasonH** [comment]
> Why the hesitancy to add something that doesn’t affect existing APIs?

Ongoing support burden in perpetuity. When we add something to the API, we’re comitting to supporting and testing it, and possibly improving it in the future. Users who may not have cared about knowing the size ahead of time might decide to care because we make it “easy” and then we might find bugs in the logic that require outsized effort to correctly support.

I’m unclear on the motivating use case, and as we’ve learned the logic is not straightforward in every case. In the meantime you can buffer the write if you need to, or fork the code.

**Comment by deitch** [comment]
> I'm not sure that we can accurately know the resulting size of the tarball before without writing out the whole thing... can we? Unless you just care about the file sizes within the tarball and not the tarball itself? What's your ultimate goal?

Actually, both the file contents inside the tarball and the tarball size itself, and you raise a good point. If I have a tarball consisting of 5 files of exactly 10KB each, then the tar file will be slightly larger than 10KB, as each file has a tar header, plus potentially some padding.

Given the complexity this creates, I am starting to see @ImJasonH's point about "ongoing support burden". I still would really like an API that returns the expected size, but understand that it just might not be realistic.

Could we at least expose the intended manifest file? This way we commit to _what_ will be in the tar file (all of the layers, the config file, and the manifest), and if a consumer wants, they can retrieve the sizes from the manifest and calculate total themselves, and with exposing the manifest, they can add that. As long as we use the same manifest inside the tar, they inherit any changes, and we need to generate and use that manifest anyways.

It also has the distinct advantage of letting consumers see what would be in the tar file once written, or even after.

I was thinking something like this. Instead of:

```go
	var m Manifest
// ...
		// Generate the tar descriptor and write it.
		m = append(m, Descriptor{
			Config:       cfgName.String(),
			RepoTags:     tags,
			Layers:       layerFiles,
			LayerSources: layerSources,
		})
// ...
	mBytes, err := json.Marshal(m)
```

We have:

```go
	var m Manifest
// ...
        m = append(m, img.TarManifest())
// ...
	mBytes, err := json.Marshal(m)
```

OK, my naming proposal may not be great, and we might want it not to be a member func, instead something like:

```go
 m = append(m, v1tarball.TarManifest(img))
```

But the idea is the same. Anyone who wants to convert it to json can do that on their own.

**Comment by jonjohnsonjr** [comment]
I'm not 100% opposed to exposing some way to compute the tar manifest, but I would like to see how it affects the code before agreeing to it. I'm just not sure if the utility outweighs the maintenance burden.

**Comment by deitch** [comment]
Yup. I will open a quick PR to show how it would be done.

**Comment by deitch** [comment]
@jonjohnsonjr see #685 for an idea of it. Open to thoughts.

**Comment by deitch** [comment]
Closed via #685
